compiler/semantic/analyzer.go

```diff
@@ -49,26 +58,26 @@ func newAnalyzer(ctx context.Context, source *data.Source, head *lakeparse.Commi
 	}
 }
 
-func (a *analyzer) enterScope() {
-	a.scope = NewScope(a.scope)
-}
-
-func (a *analyzer) exitScope() {
-	a.scope = a.scope.parent
-}
-
-func (a *analyzer) buildFrom(op dag.Op) (dag.Op, error) {
-	switch op := op.(type) {
+func HasSource(s dag.Seq) bool {
```

Comment:
Nit: We call a dag.Seq parameter `seq` everywhere else.
```suggestion
func HasSource(seq dag.Seq) bool {
```